### DISCUSSION TITLE
How can Memtest86+ remap itself during test, while the proprietary Memtest86 cannot?

### DISCUSSION THREAD

**Discussion by hap2001gh** [memtest86plus/memtest86plus#452] [General]
The proprietary version website says "MemTest86 is also inherently limited by the UEFI environment and as such, gives rise to the following limitations:......MemTest86 cannot remap itself to different portions of memory in order to run tests in the section of memory it was occupying."

But our open-source version is also 64-bit and UEFI boot compatible, while still can remap itself? Is that function actually not "limited by the UEFI environment"?

**Comment by debrouxl** [comment]
memtest86+ exits the so-called UEFI BIOS boot services (look for occurrences of `exit_boot_services` in `boot/efisetup.c`), when it does not require them anymore. In general, this operation makes it possible for the kernel to reuse for its own purposes the up to several hundreds of MBs of RAM reserved by the UEFI BIOS implementation; in the case of memtest86+, that means testing a wider range of memory, which is the essential purpose of a memory testing program.

The flip side of the coin is that after exiting the boot services, a kernel is entirely on its own: it can no longer use the UEFI BIOS capabilities for e.g. accessing partitions and filesystems on internal mass storage devices or external USB Mass Storage Devices, let alone the NIC to communicate on the network. Sometimes it's a good thing not to rely on the UEFI BIOS anymore - the SMP support in the UEFI BIOS implementation of some MBs is known to be buggy - but that represents some implementation work, of course. memtest86+ doesn't yet care about what a USB MSD and a FAT32 filesystems are and what to do with them: opening files, reading configuration, outputting reports, etc.

**Reply by changpenghui616** [reply]
I have a question， What could be the problem if not use ‘exit_boot_services’ in efi_setup.c?
im try to debug memtest86+ with debug_memtest.sh, delete ‘exit_boot_services’ , the program will crash before pem_init()， but that time not use the uefi ram of reserved yet, so whats reason
